subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -66,41 +72,46 @@ static struct lwm2m_engine_obj_field fields[] = {
 	OBJ_FIELD_DATA(FIRMWARE_UPDATE_DELIV_METHOD_ID, R, U8)
 };
 
-static struct lwm2m_engine_obj_inst inst;
-static struct lwm2m_engine_res res[FIRMWARE_MAX_ID];
-static struct lwm2m_engine_res_inst res_inst[RESOURCE_INSTANCE_COUNT];
+static struct lwm2m_engine_obj_inst inst[MAX_INSTANCE_COUNT];
+static struct lwm2m_engine_res res[MAX_INSTANCE_COUNT][FIRMWARE_MAX_ID];
+static struct lwm2m_engine_res_inst res_inst[MAX_INSTANCE_COUNT][RESOURCE_INSTANCE_COUNT];
 
 static lwm2m_engine_set_data_cb_t write_cb;
 static lwm2m_engine_execute_cb_t update_cb;
```

Comment:
If we support multiple instances now, I think it'd make sense to register write/update callbacks at per-instance basis? I guess a different handler would be used on the application side to perform applicaiton or modem FW updates?

Reply:
True, it makes sense. What do you think, would it need also backwards compatibility from the public API, as it would also add new parameters for setters and getters for the callbacks?

Reply:
I think it'd make sense to keep the existing functions, at least until the multi-instance FW update is not officially supported by the standard. The already proposed solution in `lwm2m_engine.h` for state/result setters seems fine to me, could be applied for callbacks as well (although perhaps the existing functions could be made `static inline` functions defined in the header since they'd just be thin wrappers).

Reply:
Done